commands/create.go

```diff
@@ -530,14 +546,21 @@ func updateUserdataFile(driverOpts *rpcdriver.RPCFlags, userdataFlag, customInst
 
 // replaceUserdataFile adds the customScriptContent to the user-provided userdata file and creates a new
 // temp file for this content.
+// If the user-provided userdata is empty, then we only add customScriptContent to the file.
 // If the user-provided userdata file starts with a shebang, then we can add the customScriptContent to this file.
 // fi the user-provided userdata file is in cloud-config format, then we add the customScriptContent to the `runcmd` directive.
-func replaceUserdataFile(userdataContent, customScriptContent []byte, newUserDataFile *os.File) error {
+func replaceUserdataFile(userdataContent, customScriptContent, modifiedCustomScriptContent []byte, newUserDataFile *os.File) error {
 	switch {
 	case bytes.HasPrefix(userdataContent, []byte("#!")):
 		// The user provided a script file, so the customInstallScript contents is appended to this file
-		// and written to a new file.
-		_, err := newUserDataFile.Write(bytes.Join([][]byte{userdataContent, customScriptContent}, []byte("\n\n")))
+		// and written to a new file. gzip and base64 encode the contents so it doesn't exceed user-data size limit.
+
+		encodedCustomInstallScript, err := gzipEncode(modifiedCustomScriptContent)
+		if err != nil {
+			return err
+		}
+
+		_, err = newUserDataFile.Write(bytes.Join([][]byte{userdataContent, []byte(encodedCustomInstallScript)}, []byte("\n\n")))
```

Comment:
I think the easiest thing to do here would be to use the `addToCloudConfig` with `write_files` (L589) and `runcmd` (L594) to add both the user's userdata file and the completely encoded version of the custom install script. In this case, you would encode `customScriptContent` instead of `modifiedCustomScriptContent`.

That way, the userdata contents is always provided in cloud-config format.